plugins/inputs/modbus/modbus_test.go

```diff
@@ -614,144 +625,244 @@ func TestHoldingRegisters(t *testing.T) {
+	require.NoError(t, handler.Connect())
 	defer handler.Close()
-	client := m.NewClient(handler)
+	client := mb.NewClient(handler)
 
 	for _, hrt := range holdingRegisterTests {
 		t.Run(hrt.name, func(t *testing.T) {
-			_, err = client.WriteMultipleRegisters(hrt.address[0], hrt.quantity, hrt.write)
-			assert.NoError(t, err)
+			_, err := client.WriteMultipleRegisters(hrt.address[0], hrt.quantity, hrt.write)
+			require.NoError(t, err)
 
 			modbus := Modbus{
 				Name:       "TestHoldingRegisters",
 				Controller: "tcp://localhost:1502",
-				SlaveID:    1,
-				HoldingRegisters: []fieldContainer{
-					{
-						Name:      hrt.name,
-						ByteOrder: hrt.byteOrder,
-						DataType:  hrt.dataType,
-						Scale:     hrt.scale,
-						Address:   hrt.address,
-					},
+				Log:        testutil.Logger{},
+			}
+			modbus.SlaveID = 1
+			modbus.HoldingRegisters = []fieldDefinition{
+				{
+					Name:      hrt.name,
+					ByteOrder: hrt.byteOrder,
+					DataType:  hrt.dataType,
+					Scale:     hrt.scale,
+					Address:   hrt.address,
 				},
-				Log: testutil.Logger{},
 			}
 
-			err = modbus.Init()
-			assert.NoError(t, err)
+			expected := []telegraf.Metric{
+				testutil.MustMetric(
+					"modbus",
+					map[string]string{
+						"type":     cHoldingRegisters,
+						"slave_id": strconv.Itoa(int(modbus.SlaveID)),
+						"name":     modbus.Name,
+					},
+					map[string]interface{}{hrt.name: hrt.read},
+					time.Unix(0, 0),
+				),
+			}
+
 			var acc testutil.Accumulator
-			assert.NoError(t, modbus.Gather(&acc))
-			assert.NotEmpty(t, modbus.registers)
+			require.NoError(t, modbus.Init())
+			require.NotEmpty(t, modbus.requests)
+			require.NoError(t, modbus.Gather(&acc))
+			acc.Wait(len(expected))
 
-			for _, coil := range modbus.registers {
-				assert.Equal(t, hrt.read, coil.Fields[0].value)
-			}
+			testutil.RequireMetricsEqual(t, expected, acc.GetTelegrafMetrics(), testutil.IgnoreTime())
 		})
 	}
 }
 
 func TestReadMultipleCoilLimit(t *testing.T) {
 	serv := mbserver.NewServer()
-	err := serv.ListenTCP("localhost:1502")
-	assert.NoError(t, err)
+	require.NoError(t, serv.ListenTCP("localhost:1502"))
 	defer serv.Close()
 
-	handler := m.NewTCPClientHandler("localhost:1502")
-	err = handler.Connect()
-	assert.NoError(t, err)
+	handler := mb.NewTCPClientHandler("localhost:1502")
+	require.NoError(t, handler.Connect())
 	defer handler.Close()
-	client := m.NewClient(handler)
+	client := mb.NewClient(handler)
 
-	fcs := []fieldContainer{}
+	fcs := []fieldDefinition{}
+	expectedFields := make(map[string]interface{})
 	writeValue := uint16(0)
-	for i := 0; i <= 4000; i++ {
-		fc := fieldContainer{}
+	readValue := uint16(0)
+	for i := 0; i < 4000; i++ {
+		fc := fieldDefinition{}
 		fc.Name = fmt.Sprintf("coil-%v", i)
 		fc.Address = []uint16{uint16(i)}
 		fcs = append(fcs, fc)
 
-		t.Run(fc.Name, func(t *testing.T) {
-			_, err = client.WriteSingleCoil(fc.Address[0], writeValue)
-			assert.NoError(t, err)
-		})
```

Comment:
removing the t.Run here changes the functionality, instead I think the tests will stop on the first failure instead of reporting all of them.

Reply:
Well the idea here is to write the test values into the mobus-server registers. If this fails it will fail for all values, i.e. the server doesn't work... But I can change this back if you want me to.